compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs

```diff
@@ -730,21 +756,206 @@ fn function_def_return_type(
+    let mut body_quoted = add_token_spans(body_tokens.clone(), body_argument_location.span);
+
+    // Surround the body in `{ ... }` so we can parse it as a block
+    body_quoted.0.insert(0, SpannedToken::new(Token::LeftBrace, location.span));
+    body_quoted.0.push(SpannedToken::new(Token::RightBrace, location.span));
+
+    let body =
+        parser::block(parser::fresh_statement()).parse(body_quoted).map_err(|mut errors| {
+            let error = errors.swap_remove(0);
+            let rule = "a block";
+            InterpreterError::FailedToParseMacro {
+                error,
+                tokens: body_tokens,
+                rule,
+                file: location.file,
+            }
+        })?;
+
+    let func_meta = interpreter.elaborator.interner.function_meta_mut(&func_id);
+    func_meta.has_body = true;
+    func_meta.function_body = FunctionBody::Unresolved(FunctionKind::Normal, body, location.span);
+
+    Ok(Value::Unit)
+}
+
+// fn set_parameters(self, parameters: [(Quoted, Type)])
+fn function_def_set_parameters(
+    interpreter: &mut Interpreter,
+    arguments: Vec<(Value, Location)>,
+    location: Location,
+) -> IResult<Value> {
+    let (
+        (self_argument, self_argument_location),
+        (parameters_argument, parameters_argument_location),
+    ) = check_two_arguments(arguments, location)?;
+
+    let func_id = get_function_def(self_argument, self_argument_location)?;
+    let func_meta = interpreter.elaborator.interner.function_meta(&func_id);
+    match func_meta.function_body {
+        FunctionBody::Unresolved(_, _, _) => (),
+        FunctionBody::Resolving | FunctionBody::Resolved => {
+            return Err(InterpreterError::CannotMutateFunction { location })
+        }
+    }
+
+    let (input_parameters, _type) = get_slice(
+        interpreter.elaborator.interner,
+        parameters_argument,
+        parameters_argument_location,
+    )?;
+
+    // What follows is very similar to what happens in Elaborator::define_function_meta
+    let mut parameters = Vec::new();
+    let mut parameter_types = Vec::new();
+    let mut parameter_idents = Vec::new();
+
+    for input_parameter in input_parameters {
+        let mut tuple = get_tuple(
+            interpreter.elaborator.interner,
+            input_parameter,
+            parameters_argument_location,
+        )?;
+        let parameter_type = get_type(tuple.pop().unwrap(), parameters_argument_location)?;
+        let parameter_name_tokens = get_quoted(tuple.pop().unwrap(), parameters_argument_location)?;
+        let parameter_name_quoted =
+            add_token_spans(parameter_name_tokens.clone(), parameters_argument_location.span);
+        let parameter_pattern =
+            parser::pattern().parse(parameter_name_quoted).map_err(|mut errors| {
+                let error = errors.swap_remove(0);
+                let rule = "a pattern";
+                InterpreterError::FailedToParseMacro {
```

Comment:
Looks like there's a lot of code in this loop. I wonder if we can clean it up any more?

E.g. I think:
1. Add token spans to quoted value
2. Parse token value with `[parser]`
3. Error with a `FailedToParseMacro`

Is a common pattern we can probably create a helper function for.

Reply:
Good idea, done!